planner/core/rule_column_pruning.go

```diff
@@ -309,7 +309,9 @@ func (ds *DataSource) PruneColumns(parentUsedCols []*expression.Column, opt *log
 	originSchemaColumns := ds.schema.Columns
 	originColumns := ds.Columns
 	for i := len(used) - 1; i >= 0; i-- {
-		if !used[i] && !exprUsed[i] {
+		// If ds has a shard index, it can't prune the generated column of shard index
+		// so add the new condtion `!ds.containExprPrefixUk`
+		if !used[i] && !exprUsed[i] && !ds.containExprPrefixUk {
```

Comment:
Still curious why we just prune no columns when there's the index.

Reply:
When we use `uk(tidb_shard(a), a)` , it creates a generated column like `shardfield int Generated Always as (tidb_shard(a))` implicitly. There's a sql like `SELECT * FROM t WHERE a = 10`,  it doesn't use the column `shardfield` so that `shardfield` was pruned and we can't use it later in `ds.Columns`.

Reply:
It's not about the added virtual column.
When it's a normal column and it's not used, it will still not be pruned since the `containExprPrefixUk` is true.

Reply:
Yes, I am lazy when I implement this function because of the principle of 'as simple as possible'. 
I fixed it, please check it , thank you!!